### PR TITLE
Report stored procedure errors when using info messages

### PR DESCRIPTION
**Pull request by bvogelzang**
### Summary

Capturing errors while in a non-blocking state was originally structured to capture a single error. This was intentional in order to avoid capturing more generic info messages that FreeTDS might send before the Global VM Lock was obtained. In most circumstances this is what we want. However, now that we capture info messages it is possible that a info message will be stored and the actual runtime error will be mistakenly discarded as non-important. The result is that while a runtime error is reported in the database, a `TinyTds` error is never thrown and only the info message is handled. A subset of this problem is that only one info message can be captured while in non-blocking mode which prevents stored procedures from reporting multiple info messages to `TinyTds`.

To fix this issue, the reported messages are stored within a dynamic array of `tinytds_errordata` structs, then processed normally once the GVL is obtained.

Given the fact that we don't know the number of messages that will be sent, we dynamically manage and re-allocate memory for the `nonblocking_errors` array as needed. We can't use the ruby C API because it is not safe to call while in a non-blocking state as shown by #133.

### Example

Given this procedure:

```sql
CREATE PROCEDURE tinytds_TestPrintWithError
AS
    PRINT 'hello'
    RAISERROR('Error following print', 16, 1)
```

Result:

```ruby
@client = TinyTds::Client.new({host: 'localhost', username: 'tinytds', password: '', database: 'tinytdstest', message_handler: Proc.new { |m| puts "Info: #{m.message}" }})
@client.execute("exec tinytds_TestPrintWithError").do

# Before - no error raised

Info: hello
=> -1

# After - error properly raised

Info: hello
Traceback (most recent call last):
        ...
        1: from (irb):10:in `do'
TinyTds::Error (Error following print)

```

### Notes

- The memory management could use some extra attention. Curious to hear thoughts on how this should be handled or improved.

### CODE REVIEW

**Comment by aharpervc** [comment]
This looks really good, and I also see that CI has passed, which is a good sign. I think this can probably be merged after 2.1.3 is released.

File: ext/tiny_tds/client.c
Comment (bvogelzang) [inline ext/tiny_tds/client.c:193]: Just realized that this line change might have been left over from a rebase. I'll have to take another look to see if it's needed or not.

**Comment by aharpervc** [comment]
Idea: can something be done to flush the error array, and if so, would that address this bug: https://github.com/rails-sqlserver/tiny_tds/issues/445

**Comment by bvogelzang** [comment]
I can't think of a way to adjust this in a way that would ultimately solve the timeout issue. See my commit message here: https://github.com/rails-sqlserver/tiny_tds/compare/master...bvogelzang:network-timeout-handling. 

In cases of network failure I think the problem is:

1. `dbcancel` does not succeed because of the network failure
2. `dbsqlok` does not return due to above until the underlying network timeout on the os is hit. 
3. `TinyTds` does not throw an error in the expected timeout window.

To _really_ fix that issue we'd need some way to detect the difference between a timeout error on the happy path (with normal network) vs. the bad path (with network failture). Otherwise we can't branch accordingly. It would be nice if we could detect if `dbcancel` has failed for this purpose but it looks like it always returns `SUCCEED` at the moment 😞 https://github.com/FreeTDS/freetds/blob/master/src/dblib/dblib.c#L3411.

File: ext/tiny_tds/client.c
Comment (aharpervc) [inline ext/tiny_tds/client.c:149]: This logic applies to all scenarios where we might get multiple messages, not just sp's, right?

File: ext/tiny_tds/client.c
Comment (aharpervc) [inline ext/tiny_tds/client.c:65]: This seems odd, so I have a few questions (not a pro with C, these are real questions)

1. When would `push_userdata_error` be called, and this if statement _not_ be true. I'd expect that every time we need to push an error on to the array, we'd need to resize the array.
2. Why is `nonblocking_errors_size` being doubled every time? Wouldn't it be preferable to incrementally grow the size in multiples of `sizeof(tinytds_errordata)`?
3. Why isn't the realloc like count x size, eg `realloc(userdata->nonblocking_errors, (userdata->nonblocking_errors_length + 1) * sizeof(tinytds_errordata))`?
3. Instead of tracking more state as a struct field, is anything gained by doing `sizeof(userdata->nonblocking_errors)` (assuming this would give the right answer, and if not why not?)

File: ext/tiny_tds/client.c
Comment (aharpervc) [inline ext/tiny_tds/client.c:67]: This method might get marginally simpler if `userdata->nonblocking_errors_length += 1` is added to the top of the method, then subsequent sizes and so forth are compared to that.

File: test/test_helper.rb
Comment (aharpervc) [inline test/test_helper.rb:206]: This test is specific to "message then error". We might want another test which is "message followed by more messages".

Maybe like:

```ruby
def sp_several_prints_sql
  %|CREATE or alter PROCEDURE tinytds_TestSeveralPrints
    AS
    PRINT 'hello 1'
    PRINT 'hello 2'
    PRINT 'hello 3' |
end

# ---

it 'must raise an error for each of a series of `print` messages' do
  messages.clear
  action = lambda { @client.execute("EXEC tinytds_TestSeveralPrints").do }
  assert_raise_tinytds_error(action) do |e|
    assert_equal ['hello 1', 'hello 2', 'hello 3'], messages.map { |e| e.message }, 'message list'
  end
end
```

File: ext/tiny_tds/client.h
Comment (aharpervc) [inline ext/tiny_tds/client.h:8]: What's special about 2?

**Review by aharpervc — CHANGES_REQUESTED** [review]

File: ext/tiny_tds/client.c
Comment (bvogelzang) [reply ext/tiny_tds/client.c:149]: Yup, good catch. This was due to lack of understanding when I was first writing this up.

File: ext/tiny_tds/client.c
Comment (bvogelzang) [reply ext/tiny_tds/client.c:65]: If we attempt to push an error to the array and we don't have any more allocated space to add the new error we double the allocated size for the array. This allocates space for the new error as well as _future_ space that we may end up using. This is pretty standard practice for creating dynamic arrays. The reason being that reallocating the new space is expensive (`O(n)`) since the memory may have to be copied somewhere else in order to find the contiguous space for the new array size. As a result we want to reduce the number of times we have to do that operation.

1. `nonblocking_errors_size` represents the current _allocated_ size of the array. So these values are only ever the same if we've used up all of the currently allocated space.
2. The cost of memory allocation. See above.
3. The cost of memory allocation. See above.
4. `sizeof(userdata->nonblocking_errors)` would report the size of the pointer itself _not_ the size of the array that the pointer points to. So it wouldn't give the answer you're expecting.

File: ext/tiny_tds/client.h
Comment (bvogelzang) [reply ext/tiny_tds/client.h:8]: Nothing really. I picked a value that wasn't too greedy and that would support most use cases (0-2 errors reported) without needing to reallocate any memory. I'd be open to changing this to something else.

File: ext/tiny_tds/client.c
Comment (bvogelzang) [reply ext/tiny_tds/client.c:180]: We can't free the memory here if we're freeing the memory in `nogvl_cleanup`. When this is called there is nothing to free. I am open to adding `userdata->nonblocking_errors_size = 0` here for consistency or going the other route and removing the reset for `userdata->nonblocking_errors_length`.

I decided on initializing/freeing the array in `nogvl_setup` and `nogvl_cleanup` because that marks the start and end of the time in which we need to make use of the errors array. So the array is around just long enough for its purpose. If we want the errors array to have a longer life cycle I'd be open to discussing that. For example, now that we are capturing multiple errors we may want to add a way to access those errors after we raise the exception (see #78). In that case it might be worth keeping the array around longer.

Comment (aharpervc) [reply ext/tiny_tds/client.c:180]: >  I am open to adding userdata->nonblocking_errors_size = 0 here for consistency or going the other route and removing the reset for userdata->nonblocking_errors_length

I think the consistency would be preferable, as well as possibly a code comment as to why they're being zero'd but not free'd.

> I decided on initializing/freeing the array in nogvl_setup and nogvl_cleanup because that marks the start and end of the time in which we need to make use of the errors array. So the array is around just long enough for its purpose. 

Makes sense, thanks

File: ext/tiny_tds/client.c
Comment (aharpervc) [reply ext/tiny_tds/client.c:65]: Alright, thanks for explaining. I guess it makes sense to optimize for minimal number of allocation _operations_ rather than _total memory size_, as a strategy. I will defer to your judgement on that this is both the right strategy and the appropriate implementation of that strategy.

File: ext/tiny_tds/client.h
Comment (aharpervc) [reply ext/tiny_tds/client.h:8]: Makes sense

File: ext/tiny_tds/client.c
Comment (bvogelzang) [reply ext/tiny_tds/client.c:67]: I opted for incrementing at the end to prevent the need for `userdata->nonblocking_errors_length - 1` in other places. 

```suggestion
  userdata->nonblocking_errors[userdata->nonblocking_errors_length] = error;
  userdata->nonblocking_errors_length++;
```

**Comment by bvogelzang** [comment]
@metaskills, @larskanis, @coderjoe, @wpolicarpo any chance for getting a review on this PR, https://github.com/rails-sqlserver/tiny_tds/pull/483, and https://github.com/rails-sqlserver/tiny_tds/pull/481

**Review by aharpervc — APPROVED** [review]

**Comment by bvogelzang** [comment]
These changes are now available to test in [2.1.4.pre](https://rubygems.org/gems/tiny_tds)

**Comment by aharpervc** [comment]
I'm testing this with `2.1.4.pre` and I'm still not seeing both message from your example sp. I get the initial `hello`, but then no subsequent error. If I decrease the severity to a message, I still only see `hello`.

**Comment by aharpervc** [comment]
While testing https://github.com/rails-sqlserver/tiny_tds/pull/481 I noticed that only `raiserror ... with nowait` seems to be displayed when there's a timeout, while `print` and `raiserror` without nowait didn't. 

- is this expected?
- can we figure out how to flush whatever buffer has these messages when there's an error?
- should we have tests for this situation?

example script (basically the same as the other PR):

```ruby
require 'tiny_tds'

def w(*args)
  puts [Thread.current.object_id, *args].join ' '
end

t = Thread.new do
  i = 0
  loop do
    i += 1
    w i
    sleep(1)
  end
end

c = TinyTds::Client.new({
  host: 'localhost',
  username: 'sa',
  password: 'Testing123@@',

  timeout: 5,
  login_timeout: 5,

  message_handler: ->(e) { w e },
})

w TinyTds::VERSION

sql = <<~SQL
create or alter procedure tinytds_TestPrintWithError
as
begin
    print 'print 1' -- nope
    raiserror('raiserror 1', 1, 1) -- nope
    -- raiserror('raiserror 2', 1, 1) with nowait -- yep, and uncommenting this will show all 3 messages

    waitfor delay '00:01:00'
end
SQL

w c.execute(sql).to_a
w c.execute('exec tinytds_TestPrintWithError').to_a

t.exit
```

`raiserror with nowait` commented out: 

```
$ bundle exec ruby .\scratch.rb
24279960 1
18084080 2.1.4.pre
18084080
24279960 2
24279960 3
24279960 4
24279960 5
24279960 6
Traceback (most recent call last):
        2: from ./scratch.rb:42:in `<main>'
        1: from ./scratch.rb:42:in `to_a'
./scratch.rb:42:in `each': Adaptive Server connection timed out (TinyTds::Error)
```

uncommented:

```
$ bundle exec ruby .\scratch.rb
28287960 1
18149600 2.1.4.pre
18149600
28287960 2
28287960 3
28287960 4
28287960 5
28287960 6
18149600 print 1
18149600 raiserror 1
18149600 raiserror 2
Traceback (most recent call last):
        2: from ./scratch.rb:42:in `<main>'
        1: from ./scratch.rb:42:in `to_a'
./scratch.rb:42:in `each': Adaptive Server connection timed out (TinyTds::Error)
```

**Comment by bvogelzang** [comment]
> is this expected?

I'm not exactly sure. This really just comes down to the order of _when_ freetds notifies tiny_tds of the info message. With `nowait` in this scenario freetds calls the dbmsghandle before the dberrhandle. Without it, freetds calls the dbmsghandle after calling the dberrhandle. It's unclear to me if this is what _should_ be happening or if this is a ordering bug within freetds.

> can we figure out how to flush whatever buffer has these messages when there's an error?

Yes, we can definitely do this. These info messages are being pushed to our nonblocking_errors array which means we have access to them in `novgl_cleanup`. We just need to look past any errors in the array for any remaining info messages. e.g

```c
for (short int i = 0; i < userdata->nonblocking_errors_length; i++) {
    tinytds_errordata error = userdata->nonblocking_errors[i];

    // look ahead to drain any info messages ahead of raising an error
    if (!error.is_message) {
      for (short int j = i; j < userdata->nonblocking_errors_length; j++) {
        tinytds_errordata msg_error = userdata->nonblocking_errors[j];
        if (msg_error.is_message) {
          rb_tinytds_raise_error(client,
            msg_error.is_message,
            msg_error.cancel,
            msg_error.error,
            msg_error.source,
            msg_error.severity,
            msg_error.dberr,
            msg_error.oserr
          );
        }
      }
    }

    rb_tinytds_raise_error(client,
      error.is_message,
      error.cancel,
      error.error,
      error.source,
      error.severity,
      error.dberr,
      error.oserr
    );
  }
```
I'm trying to come up with a reason we _wouldn't_ want to do this but can't really think of a downside.

> should we have tests for this situation?

Yes, we probably should.

File: CHANGELOG.md
Comment (aharpervc) [inline CHANGELOG.md:3]: "preceded"

**Comment by aharpervc** [comment]
These changes look good to me. We've been testing 2.1.4-pre2 in production for a week or so and this change seems to have the intended outcome.

Can you squash/edit your commits down to whatever's appropriate for master history before merging?

**Pull request merged**
